net/httpclient.go

```diff
@@ -207,6 +207,11 @@ type Options struct {
 
 	// Log is used for error logging
 	Log logging.Logger
+
+	// BeforeSend is a hook function that runs just before executing RoundTrip(*http.Request)
+	BeforeSend func(opentracing.Span, *http.Request)
+	// AfterResponse is a hook function that runs just after executing RoundTrip(*http.Request)
+	AfterResponse func(opentracing.Span, *http.Response, error)
```

Comment:
Request context already contains the span so there is no need to pass it as an argument.
It also ties API to opentracing. To pass something else than span along the request we would need to put it into requets context anyways.

The same thing could be implemented by wrapping http.Transport into custom roundTripper and passing it in the Options (we may need to change Options.Transport to Options.RoundTripper (or add Options.RoundTripper))

Reply:
> Request context already contains the span so there is no need to pass it as an argument. It also ties API to opentracing. To pass something else than span along the request we would need to put it into requets context anyways.

1. http.Response has no access to the Span, so it makes IMO more sense to have it analog in Before / After case.
2. simplicity for the user
3. why should we pack/unpack twice?

I agree that opentracing.Span is not ideal, but your choice was to stay with opentracing and not to move to OTel... 

> The same thing could be implemented by wrapping http.Transport into custom roundTripper and passing it in the Options (we may need to change Options.Transport to Options.RoundTripper (or add Options.RoundTripper))

Of course you can always wrap it but this should be the rare case and not the general case. The idea was to have a client that has all these batteries included and fix all stdlib bugs we know which would take long to fix in stdlib.

Reply:
> http.Response has no access to the Span

Response holds a link to the request, You can also pass request+response to AfterResponse

Reply:
this was done